### PR TITLE
New Feature Status Screen for Monitoring Temperatues with live icons …

### PR DESCRIPTION
**Pull request by guruathwal**
…**need testing on TFT24/TFT28**

New Feature Status Screen for Monitoring Temperatues with live icons.

⚠️  **only tested on TFT35 V2, needs testing on TFT24/TFT28**

👉  Currently Status Screen will only be visible if Unified Menu is activated as there was not empty position in regular home screen for Back button, Regular home screen need change in layout for this.

👉 may contain bugs so please be calm and open an issue. :+1:

~~Best viewed on printer with dual extruder setup~~
the screen will now shuffle between multiple extruders and fan if they enabled. 

photo_2019-11-13_22-24-13

### CODE REVIEW

**Comment by Hukuma1** [comment]
Post a photo please! :) Would love to see it.

**Comment by hapklaar** [comment]
Nice. Will this also run on a TFT35 v3? If so I can test.

**Comment by guruathwal** [comment]
@hapklaar yes you can test it.
@Hukuma1 I will post a ptoho for sure.

**Comment by mooner1022** [comment]
Sounds good!

**Comment by guruathwal** [comment]
Added Speed and Flow icons.
IMG_20191112_222539708

Only tested on TFT35 V2

Needs Testing on TFT24 / TFT28 and TFT35 V1 and V3.

**Comment by thisiskeithb** [comment]
I’ll be able to test on a TFT35 V3 later today.

**Comment by hapklaar** [comment]
Initial findings: Love where this is going :) Seems to work without any issue on TFT35 v3

Is it possible to use the unused space on the status screen to show M117 message? That way you can display custom status/progress info from for example [OctoPrint-DetailedProgress](https://github.com/jneilliii/OctoPrint-DetailedProgress)

**Comment by Zuru1** [comment]
> 
> 
> Added Speed and Flow icons.
> IMG_20191112_222539708
> 
> Only tested on TFT35 V2
> 
> Needs Testing on TFT24 / TFT28 and TFT35 V1 and V3.

Nice work 👍 
One question, is it possible to add a second extruder?

**Comment by guruathwal** [comment]
> One question, is it possible to add a second extruder?

@Zuru1 yes yo can add upto 6 extruders. The screen will cycle between each extruder every 2 second. Same goes for the Fan.

@hapklaar  yes it is possible. Working on next PR.

**Comment by guruathwal** [comment]
@hapklaar is this what you requested?

photo_2019-11-13_22-24-13

Here is a video demo:
https://youtu.be/MmQEkVuC5to

**Comment by thisiskeithb** [comment]
BigTreeTech might have to make this the default layout.

**Comment by hapklaar** [comment]
> @hapklaar is this what you requested?

It is, exactly what I had in mind :) Great work man, thanks! I will try this tomorrow or in the weekend.

**Comment by zeko00** [comment]
how can I activate te status screen to test it

**Comment by guruathwal** [comment]
photo_2019-11-13_22-24-13



Pushed PR for the information box will now show Echo messages from the printer and M117 command message from ESP3D when status screen is active, on any other screen it will show regular pop up window.

**Comment by guruathwal** [comment]
@zeko00 currently you can use this fork to test it: https://github.com/guruathwal/BIGTREETECH-TouchScreenFirmware as the PR is not yet merged.

**Comment by hapklaar** [comment]
@guruathwal I can't compile your version anymore, even after complete refresh from your fork and reinstalling pio:
image

**Comment by Sthopeless** [comment]
@guruathwal I have purchased TFT24 and TFT28, so when they arrive I can help a little more with testing.

Got it working on my TFT35 v1.2 (not attached to a printer) and everything seems to be working besides the speed/flow icon inside doesn't allow me to change the `Normal` and if I change it on `Menu > Extrude` it will not reflect that change

**Comment by Sthopeless** [comment]
Unless we have multiple fans or hotends defined wouldn't it look better if removed the text from the icons and center the icon?

**Comment by guruathwal** [comment]
@hapklaar  I tried building for all variants. only TFT35_V3_0 and TFT24_V1_1 build failed. 
it looks like there is an issuse with the Stack size in ldscripts for both of them.

I was able to success fully build the firmware for these variant by reducing the `_Min_Stack_Size` from `0x2000` to `0x1800` in following files:

**For TFT35_V3_0**
`BIGTREETECH-TouchScreenFirmware\buildroot\ldscripts\buildroot\ldscripts\stm32f2xx_0x8000_iap.ld`

https://github.com/guruathwal/BIGTREETECH-TouchScreenFirmware/blob/9b77422e849fb12ae781339e1842c8db7b8f6436/buildroot/ldscripts/stm32f2xx_0x8000_iap.ld#L15

**For  TFT24_V1_1**
`BIGTREETECH-TouchScreenFirmware\buildroot\ldscripts\buildroot\ldscripts\stm32f10x_0x6000_iap.ld`

https://github.com/guruathwal/BIGTREETECH-TouchScreenFirmware/blob/9b77422e849fb12ae781339e1842c8db7b8f6436/buildroot/ldscripts/stm32f10x_0x6000_iap.ld#L40

**Comment by guruathwal** [comment]
> @guruathwal I have purchased TFT24 and TFT28, so when they arrive I can help a little more with testing.
> 
> Got it working on my TFT35 v1.2 (not attached to a printer) and everything seems to be working besides the speed/flow icon inside doesn't allow me to change the `Normal` and if I change it on `Menu > Extrude` it will not reflect that change

@Sthopeless Thank you for your help.
The function of  `Normal` button in `Menu > Extrude` is different from the `Normal` button in speed/flow setting.
in Extrude menu `Normal` button changes the speed of extrude but in Speed/Flow the `Normal` button is used to reset the speed/flow setting to 100% (default).
it is a bit confusing as the `Normal` button in Speed/Flow should be renamed to `Reset` which will make more sense.

**Comment by guruathwal** [comment]
> Unless we have multiple fans or hotends defined wouldn't it look better if removed the text from the icons and center the icon?

the icon are in the form of BMP image. they cannot be centered without modifying the BMP file.
however you can change the display name for each of them for example: `F0` can be changed to `Fan`.

**Comment by Sthopeless** [comment]
@guruathwal ohhhhh, now I see! yes reset makes more sense IMO

**Comment by hapklaar** [comment]
@guruathwal It looks great, but I can't get it to display any M117 messages, which are displayed ok in emulator mode. It only reads 'Ready'. Anything wrong on my end?

**Comment by Sthopeless** [comment]
minimal_ui

**Comment by guruathwal** [comment]
@hapklaar does it show any pop up screen with the Message or nothing happens at all?

**Comment by guruathwal** [comment]
@hapklaar  I tried sending M117 command from my host and it works fine. can you open the gcode terminal in your TFT and see what text is it receiving. as I see from your first request the TFT was also receiving "Connected" before the actual M117 command. can you confirm?

**Comment by vnique08** [comment]
Just wanted to add my worthless two cents. This looks awesome! Can’t wait for the final outcome!

**Comment by hapklaar** [comment]
> @hapklaar I tried sending M117 command from my host and it works fine. can you open the gcode terminal in your TFT and see what text is it receiving. as I see from your first request the TFT was also receiving "Connected" before the actual M117 command. can you confirm?

@guruathwal I don't see any output for any commands I give through Octoprints terminal in the gcode monitor on the TFT, just the M105 command and temp replies streaming by.

**Comment by guruathwal** [comment]
> @guruathwal I don't see any output for any commands I give through Octoprints terminal in the gcode monitor on the TFT, just the M105 command and temp replies streaming by.

@hapklaar try sending M117 command manually from Octoprints terminal when Status Screen is visible. and then try when any other menu is opened other that gcode monitor. share the result.

**Comment by arminth** [comment]
Can't compile dor 2.4". Compiler Error even with 1800 Stacksize.
`Compiling .pio\build\BIGTREE_TFT24_V1_1\src\src\User\Menu\menu.o
TFT\src\User\Menu\StatusScreen.c:103:138: error: 'STATUS_MSGBOX_BODY_YOFFSET' undeclared here (not in a function); did you mean 'STATUS_MSGBOX_BODY_XOFFSET'?
         GUI_RECT msgRect ={START_X + 1 * ICON_WIDTH + 1 * SPACE_X,                        TITLE_END_Y +  1 * ICON_HEIGHT + 1 * SPACE_Y + STATUS_MSGBOX_BODY_YOFFSET,
                                                                                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
                    `


[platformio.ini.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3854048/platformio.ini.txt)

[Configuration.h.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3854049/Configuration.h.txt)

**Comment by Sthopeless** [comment]
@guruathwal have you seen this suggestion https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/157 from @Hukuma1?
Maybe having this showed at the top would free up some space on the main screen for more things.

Also for the `Nozzle `and `Heated Bed` icons wouldn't it be better when pressed if they go directly to the 'temp control menu', nozzle goes directly to the nozzle and heated bed directly to the heated bed. 
And maybe for now the rest can stay at `Heat/Fan` as it is for Preheating

**Comment by guruathwal** [comment]
> Can't compile for 2.4". Compiler Error even with 1800 Stacksize.
> `Compiling .pio\build\BIGTREE_TFT24_V1_1\src\src\User\Menu\menu.o TFT\src\User\Menu\StatusScreen.c:103:138: error: 'STATUS_MSGBOX_BODY_YOFFSET' undeclared here (not in a function); did you mean 'STATUS_MSGBOX_BODY_XOFFSET'? GUI_RECT msgRect ={START_X + 1 * ICON_WIDTH + 1 * SPACE_X, TITLE_END_Y + 1 * ICON_HEIGHT + 1 * SPACE_Y + STATUS_MSGBOX_BODY_YOFFSET, ^~~~~~~~~~~~~~~~~~~~~~~~~~ `
> 
> [platformio.ini.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3854048/platformio.ini.txt)
> 
> [Configuration.h.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3854049/Configuration.h.txt)

@arminth  in the last PR 'STATUS_MSGBOX_BODY_YOFFSET' was missed
please add this line `#define STATUS_MSG_BODY_YOFFSET     45` after line number `43` in `StatusScreen.h`

https://github.com/guruathwal/BIGTREETECH-TouchScreenFirmware/blob/9b77422e849fb12ae781339e1842c8db7b8f6436/TFT/src/User/Menu/StatusScreen.h#L43

This will be fixed in next PR later today. Thanks.

**Comment by arminth** [comment]
@guruathwal Thanks a lot!
But you probably mean `#define STATUS_MSGBOX_BODY_YOFFSET 45` . That did the trick!
I also found a missing icon: setup->machine->custom->Home is not defined.

I also attached an updated German Language file.
[language_de.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3854197/language_de.zip)

**Comment by guruathwal** [comment]
> @guruathwal have you seen this suggestion #157 from @Hukuma1?
> Maybe having this showed at the top would free up some space on the main screen for more things.
> 
> Also for the `Nozzle `and `Heated Bed` icons wouldn't it be better when pressed if they go directly to the 'temp control menu', nozzle goes directly to the nozzle and heated bed directly to the heated bed.
> And maybe for now the rest can stay at `Heat/Fan` as it is for Preheating

👍 yes I have seen it. that area will also be used in future. 
Nozzle and heated bed will go directly to their respective heat menu, presently main goal is to complete the status screen.

**Comment by guruathwal** [comment]
> @guruathwal Thanks a lot!
> But you probably mean `#define STATUS_MSGBOX_BODY_YOFFSET 45` . That did the trick!
> I also found a missing icon: setup->machine->custom->Home is not defined.
> 
> I also attached an updated German Language file.
> [language_de.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3854197/language_de.zip)

👍 Anytime.
thanks for pointing out. The PR for Custom menu and German language was added after this PR. will add them also. 👍

**Comment by hapklaar** [comment]
> > @guruathwal I don't see any output for any commands I give through Octoprints terminal in the gcode monitor on the TFT, just the M105 command and temp replies streaming by.
> 
> @hapklaar try sending M117 command like `M117 test message from octoprint` manually from Octoprints terminal when Status Screen is visible. and then try when any other menu is opened other that gcode monitor. share the result.

Nothing happens on the status screen, and no window or screen pops up on other menu's. It's like the TFT doesn't receive or process these messages in TFT mode. It does receive them in Marlin mode.

**Comment by guruathwal** [comment]
@hapklaar I do not have an ESP right now to check it. can you connect ESP to PC and check the commands it sends in serial monitor and share the full text received?

**Comment by kkaleta** [comment]
I'm unable to compile firmware. Still got the same error: 
 stm32f10x_0x6000_iap.ld: No such file or directory

Am I missing something?

2019-11-17 10_14_32-Window
[stm32f10x_0x6000_iap.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3855422/stm32f10x_0x6000_iap.txt)

[platformio.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3855424/platformio.txt)

**Comment by hapklaar** [comment]
> @hapklaar I do not have an ESP right now to check it. can you connect ESP to PC and check the commands it sends in serial monitor and share the full text received?

@guruathwal I connected an ESP01 with the WiFiTelnetToSerial example sketch. No output from any commands given in Octoprint are shown on the serial monitor, I only see temp status messages
image

**Comment by guruathwal** [comment]
> @guruathwal I connected an ESP01 with the WiFiTelnetToSerial example sketch. No output from any commands given in Octoprint are shown on the serial monitor, I only see temp status messages
> 
> PS during the period in the screenshot I spammed Octoprint with M177 messages
> PS2 Noticed some very intermittent behavior, where at seemingly random intervals commands given through TFT are not received or processed by the printer anymore. Have to powercycle the lot to get it working again. Not sure what could cause this, might be an issue with config, SKR1.3 or TFT but it might be related to this particular issue.

The WiFiTelnetToSerial Sketch does not work properly with BTT TFT's. the TFT will hang most of the time.

You specified `M177 messages` are you sure you used `M177`. the correct command is `M117`
I connected the USB Serial to the ESP port on the TFT and sent the M117 command and it is working correctly.

**Comment by guruathwal** [comment]
> I'm unable to compile firmware. Still got the same error:
> stm32f10x_0x6000_iap.ld: No such file or directory
> 
> Am I missing something?
> 
> 2019-11-17 10_14_32-Window
> [stm32f10x_0x6000_iap.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3855422/stm32f10x_0x6000_iap.txt)
> 
> [platformio.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3855424/platformio.txt)

@kkaleta the `stm32f10x_0x6000_iap.txt` you attached should be named as `stm32f10x_0x6000_iap.ld`
Remove the `.txt` extension from the name and add `.ld` extension.

**Comment by hapklaar** [comment]
M177 was a typo, I used M117

**Comment by kkaleta** [comment]
> > I'm unable to compile firmware. Still got the same error:
> > stm32f10x_0x6000_iap.ld: No such file or directory
> > Am I missing something?
> > 2019-11-17 10_14_32-Window
> > [stm32f10x_0x6000_iap.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3855422/stm32f10x_0x6000_iap.txt)
> > [platformio.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3855424/platformio.txt)
> 
> @kkaleta the `stm32f10x_0x6000_iap.txt` you attached should be named as `stm32f10x_0x6000_iap.ld`
> Remove the `.txt` extension from the name and add `.ld` extension.

I've changed to .txt It's one of the allowed github attachment file extension.

**Comment by guruathwal** [comment]
> > > I'm unable to compile firmware. Still got the same error:
> > > stm32f10x_0x6000_iap.ld: No such file or directory
> > > Am I missing something?
> > > [stm32f10x_0x6000_iap.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3855422/stm32f10x_0x6000_iap.txt)
> > > [platformio.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3855424/platformio.txt)
> > 
> > 
> > @kkaleta the `stm32f10x_0x6000_iap.txt` you attached should be named as `stm32f10x_0x6000_iap.ld`
> > Remove the `.txt` extension from the name and add `.ld` extension.
> 
> I've changed to .txt It's one of the allowed github attachment file extension.

@kkaleta ok. try remove the source completely and copy again form the fork.

**Comment by guruathwal** [comment]
> M177 was a typo, I used M117
> 
> I could try connecting USB serial to the ESP port by wire, but I'm not expecting this to give different results, as the gcode monitor on the TFT also doesn't show any M117 output. Any idea what could be the issue? Broken SKR or TFT?

@hapklaar .. I checked Gcode Monitor on TFT will will not show commands from ESP.
I received my ESP today will try it on my end.

**Comment by hapklaar** [comment]
> @hapklaar .. I checked Gcode Monitor on TFT will will not show commands from ESP.
> I received my ESP today will try it on my end.

Maybe I didn't quite understand you, but I was talking about ESP and Gcode monitor not showing output from M117 given through Octoprint (connected to SKR serialport -1) for me. Should this be the case?

**Comment by kkaleta** [comment]
> > > > I'm unable to compile firmware. Still got the same error:
> > > > stm32f10x_0x6000_iap.ld: No such file or directory
> > > > Am I missing something?
> > > > [stm32f10x_0x6000_iap.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3855422/stm32f10x_0x6000_iap.txt)
> > > > [platformio.txt](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/3855424/platformio.txt)
> > > 
> > > 
> > > @kkaleta the `stm32f10x_0x6000_iap.txt` you attached should be named as `stm32f10x_0x6000_iap.ld`
> > > Remove the `.txt` extension from the name and add `.ld` extension.
> > 
> > 
> > I've changed to .txt It's one of the allowed github attachment file extension.
> 
> @kkaleta ok. try remove the source completely and copy again form the fork.

I can easily compile version from 22-10-2019. 
image
But the current one (18-11-2019) gives me the error "No such file"
image

**Comment by guruathwal** [comment]
@kkaleta This looks similar to your issue: https://community.platformio.org/t/arm-none-eabi-gcc-error/6906

**Comment by guruathwal** [comment]
> > @hapklaar .. I checked Gcode Monitor on TFT will will not show commands from ESP.
> > I received my ESP today will try it on my end.
> 
> Maybe I didn't quite understand you, but I was talking about ESP and Gcode monitor not showing output from M117 given through Octoprint (connected to SKR serialport -1) for me. Should this be the case?

Did you connected Octoprint to second serial port on the SKR?

**Comment by guruathwal** [comment]
photo_2019-11-19_01-10-03

reduced the title bar height while in status screen to display the X.Y & Z positions.
can be enabled on standard menu also.
fixed stack_size error.

File: TFT/src/User/Menu/StatusScreen.c [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+
+static u32       nowTime = 0;
+static u32   update_time = 200; // 1 seconds is 100
+SCROLL   msgScroll;
+static int lastConnection_status = -1;
+
+static char* msgtitle[20];
+static char* msgbody[512];
+
+static char* msgxyz[512];
```
[the remark this answers is not in the document]
Comment (guruathwal) [reply TFT/src/User/Menu/StatusScreen.c]: Yes it should be the first one..👍🏼

**Comment by hapklaar** [comment]
> @hapklaar Did you connected Octoprint to second serial port on the SKR?

Octoprint is connected to the onboard USB port of the SKR (which is SERIAL_PORT -1 in Marlin) and enabled the second serial port by uncommenting 'SERIAL_PORT_2 0' for TFT35

**Comment by hapklaar** [comment]
> @hapklaar Did you connected Octoprint to second serial port on the SKR?

@guruathwal Octoprint is connected to the onboard USB port of the SKR (which is SERIAL_PORT -1 in Marlin) and enabled the second serial port by uncommenting 'SERIAL_PORT_2 0' for TFT35

**Comment by guruathwal** [comment]
> > @hapklaar Did you connected Octoprint to second serial port on the SKR?
> 
> @guruathwal Octoprint is connected to the onboard USB port of the SKR (which is SERIAL_PORT -1 in Marlin) and enabled the second serial port by uncommenting 'SERIAL_PORT_2 0' for TFT35

@hapklaar  both are connected to separate serial port. M117 command form Octoprint will not show in TFT.

**Comment by hapklaar** [comment]
> > > @hapklaar Did you connected Octoprint to second serial port on the SKR?
> > 
> > 
> > @guruathwal Octoprint is connected to the onboard USB port of the SKR (which is SERIAL_PORT -1 in Marlin) and enabled the second serial port by uncommenting 'SERIAL_PORT_2 0' for TFT35
> 
> @hapklaar both are connected to separate serial port. M117 command form Octoprint will not show in TFT.

@guruathwal Does this mean the result from M117 also won't reach the TFT (guess so :)). Would it be worth it to request a change from Marlin team in processing of this command? Not sure if it would even be possible...

**Comment by guruathwal** [comment]
@hapklaar  yes it is so.. you can try with the Marlin team. you can also connect octoprint to the TFT serial port at esp mount instead of the board itself.

**Comment by hapklaar** [comment]
> @hapklaar yes it is so.. you can try with the Marlin team. you can also connect octoprint to the TFT serial port at esp mount instead of the board itself.

@guruathwal Just thinking out loud here... Wouldn't it be possible to get the M117 output another way as this is already reaching the TFT, through EXP1/EXP2 in Marlin mode?

**Comment by AnHardt** [comment]
@hapklaar 
Character recognition from (moving) pixels?
Interesting concept, but unlikely to be realized.

**Comment by carl1961** [comment]
@guruathwal How do I activated Unified Menu, So far on  TFT 3.5 V3.0 every thing is working. printed from sdcard fine on TFT and LCD12864 Simulator . Thanks for your Awesome work!

**Comment by guruathwal** [comment]
> > @hapklaar yes it is so.. you can try with the Marlin team. you can also connect octoprint to the TFT serial port at esp mount instead of the board itself.
> 
> @guruathwal Just thinking out loud here... Wouldn't it be possible to get the M117 output another way as this is already reaching the TFT, through EXP1/EXP2 in Marlin mode? I imagine most people have Octoprint connected to their mainboard opposed to the ESP port on TFT.

@hapklaar  Did you tried the `M118` command?

**Comment by guruathwal** [comment]
> @guruathwal How do I activate the Unified Menu, So far on TFT 3.5 V3.0 every thing is working. printed from sdcard fine on TFT and LCD12864 Simulator . Thanks for your Awesome work!

@carl1961 👍 ... Unified menu can be activated with this line https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/blob/de205ef1e053702663d018a824f526c4eb071d87/TFT/src/User/Configuration.h#L117

**Comment by hapklaar** [comment]
> @hapklaar yes it is so.. you can try with the Marlin team. you can also connect octoprint to the TFT serial port at esp mount instead of the board itself.

@guruathwal Still playing around with this :) I now connected an ESP01 to the TFT and loaded ESP3d onto it. On Octoprint machine I used socat to create a virtual serial port that connects to the data port of ESP3d and have Octoprint connect to that vser. It works, albeit not very stable. Also I see some very strange behavior when sending an M117 command
:
Send: M117 testing123
[...]
Recv: echo:Unknown command: "M147 testing123"
Recv: ok
Recv: echo:Unknown command: "S0"
Recv: ok

Might be bad ESP3d behavior, haven't figured it out yet. Also not very keen on connecting Octoprint and TFT with wifi.

Positive side is, I also see those errors on the status screen in TFT mode.

Can I ask your setup? How did you connect everything together if you don't have Octoprint on the mainbord USB/serial port?

**Comment by carl1961** [comment]
have you looked at https://github.com/probonopd/WirelessPrinting or  even https://github.com/jeelabs/esp-link  there made for wifi access

**Comment by guruathwal** [comment]
@hapklaar  if the TFT is displaying the echo messages then it means that your TFT is working correctly.
>Send: M117 testing123
>[...]
>Recv: echo:Unknown command: "M147 testing123"
>Recv: ok
>Recv: echo:Unknown command: "S0"
>Recv: ok

it looks like there is some noise or data is missed during communication between esp and octopi.
I do not recommend using socat. instead you should directly connect the raspberry pi serial to the serial port on ESP mount if possible.
I use both PC via USB and SD card.
The ESP I received is faulty, so I cannot confirm about the it.

As @carl1961 suggested you can also try both alternative

**Comment by hapklaar** [comment]
@guruathwal The M117 that becomes M147 folowed by S0 seems a quirk of ESP3d (https://github.com/luc-github/ESP3D). This output is consistent, doesn't seem like a noise or connection related issue. Going to try a direct cable connection. How did you exactly connect this up, a USB to serial bridge, then straight to ESP connector on the back of the TFT?

@carl1961 I used [ESP3d ](https://github.com/luc-github/ESP3D )which is more or less specifically for this purpose. Will check out your suggestions.

**Comment by davehardy20** [comment]
@hapklaar I have this same issue, TFT35 v3.0 plus BTT WiFi/ESP01S with ESP3D firmware. Same odd M147 etc etc

**Comment by hapklaar** [comment]
> @hapklaar I have this same issue, TFT35 v3.0 plus BTT WiFi/ESP01S with ESP3D firmware. Same odd M147 etc etc FYI the ESP01S is plugged into the TFT35 socket

@davehardy20 I still have no idea where this M147 is coming from. Even if I use esp-link firmware instead of ESP3d to enter the M117 command, the behavior is identical with strange M147 output...

image

**Comment by carl1961** [comment]
sound like a custom CNC code, S0 is usually speed. searching M147 brings us a CNC clamp

**Comment by hapklaar** [comment]
Yeah I found that as well by text searching through his code. But still can't figure out why this is happening. Opened an issue at ESP3d hoping to find out more.

https://github.com/luc-github/ESP3D/issues/388

**Comment by carl1961** [comment]
it says it is in testui.htm.  Using UltraFinder , searching all the BIGTREETECH-TouchScreenFirmware code nothing bring up M147. 
Ultiment Finder

**Comment by hapklaar** [comment]
Just checked and the ESP module is passing M117 without issue to ESP port, please see mention above. It seems to be the TFT is doing weird things with it.

**Comment by guruathwal** [comment]
Annotation 2019-11-24 204020

photo_2019-11-24_20-43-44

I connected USB to TTL on the ESP port on TFT and sent the M117 code. it works as expected..

**Comment by hapklaar** [comment]
@guruathwal I downloaded your fork, compiled and flashed it. Seems to work fine with that one.

Please also check https://github.com/luc-github/ESP3D/issues/388 where the owner of ESP3d says he might have found the issue. I can't explain why this only happens in the original code and not in your fork though...

**Comment by guruathwal** [comment]
> @guruathwal I downloaded your fork, compiled and flashed it. Seems to work fine with that one.
> 
> Please also check [luc-github/ESP3D#388](https://github.com/luc-github/ESP3D/issues/388) where the owner of ESP3d says he might have found the issue. I can't explain why this only happens in the original code and not in your fork though...

@hapklaar 👍 may be I am lucky for you..😆

**Comment by luc-github** [comment]
FYI the bug was introduced here : https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/commit/cba9b7d18f5b4520fc88024a97974063437ba628#diff-2648696ce3fb72180bd1f1f65adfc697

**Comment by hapklaar** [comment]
@luc-github tried your fix on the original tree and it works! Shall I create a PR for this?

@guruathwal you sure are, without you we wouldn't have this great new UI design and status screen 😆

**Comment by hapklaar** [comment]
@guruathwal oh you already fixed this yourself? That's why your fork had no issue today :)

https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/commit/ca9f44b95893638ba111e4fde317d683c3711b01

**Comment by guruathwal** [comment]
@hapklaar 👍🏼 the maintainers are a litle slow with the merges..

**Comment by ssombra** [comment]
> @hapklaar yes it is so.. you can try with the Marlin team. you can also connect octoprint to the TFT serial port at esp mount instead of the board itself.

Hi, newbie here. How can I connect Octoprint to TFT serail port? I have the TFT35 v2 display, any guide?. thanks

**Comment by QuitButton** [comment]
I can't get M117 to show at all. I've just flashed with TFT24_V1_1_26 and there's nothing displayed on the screen when I send a M117 command. Do I need to enable something somewhere?

**Comment by JoachimBates** [comment]
Hi All, correct me if I’m wrong but I cannot for the life of me understand how to install this “modded” screen setup on my CR10-s pro with BTT SKR1.4 and stock creality touch display. It’s far more beautiful and user friendly than the stock Marlin Menu style 

Joe

**Comment by carl1961** [comment]
@JoachimBates  You have to have a BIGTREETECH Touch Screen Display

**Pull request merged**